### DISCUSSION TITLE
How can I get which variables I need to pass through the scope dynamically?

### DISCUSSION THREAD

**Discussion by KerusMighos** [josdejong/mathjs#3302] [Q&A]
How can I get which variables I need to pass through the scope dynamically?

Given an arbitrary expression written by a user, I want to be able to find out which "variables" this expression expects in the scope, and if possible, the type of those variables.

Some ideas I've thought of:

1. I have the user write their expression, and also write which variables their expression expects (this would be written outside the expression, in json for example)

2. I create and expose a function, something like getArg('argName', 'argType'), so instead of the expression simply trying to access a variable in an expression, I would initialize all the variables with something like arg1 = getArg('arg1', 'number').

3. I go through the ast looking for all the symbols that were used before they were assigned.

**Comment by josdejong** [comment]
Maybe the following topics help:

- https://github.com/josdejong/mathjs/discussions/2551
- https://github.com/josdejong/mathjs/discussions/2865
- https://github.com/josdejong/mathjs/issues/1414
- https://github.com/josdejong/mathjs/issues/660
